packages/backend-common/src/tokens/AzureAccessToken.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { ClientSecretCredential } from '@azure/identity';
+import { TokenCredential } from '@azure/core-auth';
+import { InputError } from '@backstage/errors';
+
+/**
+ * Helpers singtone class to distunguiesh connection.
+ * @TODO remove before merge.
+ */
+export class ConnectionCounter {
+  private static _instance: ConnectionCounter;
+  public static counter: number = 0;
+  public static get Instance() {
+    return this._instance || (this._instance = new this());
+  }
+  public static increment(): number {
+    ConnectionCounter.counter = ConnectionCounter.counter + 1;
+    return ConnectionCounter.counter;
+  }
+}
+
+/**
+ * Retrieves Azure AD access token information
+ *
+ * @public
+ */
+export class AzureAccessToken {
+  public credential: TokenCredential;
+  public counter: number;
+  constructor(
+    private tenantId: string,
+    private clientId: string,
+    private clientSecret: string,
+  ) {
```

Comment:
I did refactor the class to a single function that has  `TokenCredential` as an argument. I do not see the need for a class there anymore.
However, I didn't get the idea with fall back to a DefaultCredential. Can you elaborate on it a bit?
Thank you

Reply:
When you use, for example, a managed identity, you don't have access to this identity when your run backstage locally. So in our local dev environment, you can use your user's credentials instead. There are several ways to get your Azure Credentials, like from the Azure CLI context, Visual Studio, Visual Studio Code, etc. A [DefaultAzureCredential](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet), is a chained credential that tries different kinds of credentials in order until it finds something that works.

So if you can fall back to a `DefaultAzureCredential` when running locally, then you'd be able to use your own user's credentials to connect to the database instead of requiring a service principal or password in your `appconfig.local.yaml`

Reply:
Thank you for the explanation @sanderaernouts. I did add a fallback to [DefaultAzureCredential](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.defaultazurecredential?view=azure-dotnet).
Take a look please :)